orangecontrib/spectroscopy/widgets/owpolar.py

```diff
@@ -838,11 +838,16 @@ def commit(self):
         else:
             sorted_data = self.sorted_data
 
+        self.setInvalidated(True)
```

Comment:
This is not the way to do it. 

Just calling `self.cancel()` in the `commit` function (that calls the `start`) would be sufficient. Please try that instead.

Reply:
Hi Marko,

I originally tried;
'def cancel(self):
    super().__cancel_task(wait=True)'
in hopes that it would wait for the tasks to complete. The issue I was having with that is if the inputs were changed when the code was running in the sharedmemorymanager block, the code would continue but with empty data and so the curve fit procedure would try to fit to an empty array. I defined the function beneath 'onDeleteWidget'.

I would have though simply calling 'self.cancel()' in 'commit' would always try and stop the process?

Reply:
If you reimplemented cancel nothing would happen unless you call it. And likely this is not needed.

Adding `cancel()` to commit should ensure that the new process is stopped before running a new one if that was the only issue. 

Theoretically you could run cancel in the input handling functions, but as it seems you pass all needed data into `start_compute` and do everything within (which is the correct way to do it), the actual objects are unchanged that should be not needed. If, on the other hand, we would be passing lists that are only edited when a new input arrives and not reinitialized, the particular list withing the computation (its object) would be the same, thus edited, thus problems. So we would actually have to avoid that.

So not adding `cancel()` to the inputs should not affect crashes if the rest is written well.

For this kind of issues I'd advise first making a test case in `test_owpolar` that showcases the error and then fixing the code so that it disappears. That would also allow me to easily replicate your issue and debug it.

Reply:
I gave this another shot today. Unfortunately calling `cancel()` doesn't resolve the issue :(

I did some testing and what I found was:

- If the input is disconnected before the computation processes have started the code appears to continue until the `with` block is finished, but a ValueError (operands could not be broadcast together with shapes (0,) (4,)) is eventually thrown with the traceback originating from line 240: `params = curve_fit(azimuth, x, temp, jac=azimuth_jac)[0]`. So it looks like the data within the sharedmemory arrays are being erased... however, the allocated memory is not being cleared as seen by a steadily increasing memory usage with repeated tries.
- If the input is disconnected after the computation processes have started, no error is given, but the allocated memory is still not released.
- In both of the above cases, no code is executed after the `with` block
- The SharedMemoryManager is created within a new child process, and so to me this seems logical that code within the `with` block continues to execute.

From this, it would seem to me that this is occurring, but I am not entirely sure how...
>  If, on the other hand, we would be passing lists that are only edited when a new input arrives and not reinitialized, the particular list withing the computation (its object) would be the same, thus edited, thus problems. So we would actually have to avoid that.

I have added a "kind-of" test for this in test_owpolar. It does not result in a failure of the test, but it does show the ValueError mentioned above. As the error occurs in a child process it is not so easy to catch to result in the test failing.


I will continue to try find another work around for this, but if there are any suggestions they would be greatly appreciated!